File: src/Python/read_test.py
# Masahiro Furukawa
# June 30, 2020

import serial
import time
from struct import *

# ser = serial.Serial('/dev/ttyUSB0', 921600)  
ser = serial.Serial('COM4', 921600,  timeout=1)  

state = 0
buf = []

class ACC_GYRO:
    def __int__(self):
        self.ax = 0
        self.ay = 0
        self.az = 0
        self.wx = 0
        self.wy = 0
        self.wz = 0


# data frame    
ch = []
ch.append( ACC_GYRO() )
ch.append( ACC_GYRO() )
ch.append( ACC_GYRO() )
ch.append( ACC_GYRO() )

# command char 'r' means to stop streaming
ser.write(b'r')
time.sleep(0.01)

# command char 's' means to start streaming
ser.write(b's')
time.sleep(0.01)

while 1:

    val = ser.read()
    
    # waiting for '\r'
    if state == 0 and val == b'\r':        
        val = ser.read()
        
        # waiting for '\n' and get ready
        if val == b'\n':
            state = 1
            buf = []
            next
        else:
            print('End byte set failure.')

    # store val into list
    if state == 1:
        buf.append(val)
        
        # decode 48 bytes
        if len(buf) == 48:
            (ch[0].ax, ch[0].ay, ch[0].az, ch[0].wx, ch[0].wy, ch[0].wz,
             ch[1].ax, ch[1].ay, ch[1].az, ch[1].wx, ch[1].wy, ch[1].wz,
             ch[2].ax, ch[2].ay, ch[2].az, ch[2].wx, ch[2].wy, ch[2].wz,
             ch[3].ax, ch[3].ay, ch[3].az, ch[3].wx, ch[3].wy, ch[3].wz )= unpack('hhhhhhhhhhhhhhhhhhhhhhhh', b''.join(buf)) 
            
            print (ch[0].ax, ch[0].ay, ch[0].az, ch[0].wx, ch[0].wy, ch[0].wz,
             ch[1].ax, ch[1].ay, ch[1].az, ch[1].wx, ch[1].wy, ch[1].wz,
             ch[2].ax, ch[2].ay, ch[2].az, ch[2].wx, ch[2].wy, ch[2].wz,
             ch[3].ax, ch[3].ay, ch[3].az, ch[3].wx, ch[3].wy, ch[3].wz ) 
            

            # print(unpack('hhhhhhhhhhhhhhhhhhhhhhhh', b''.join(buf)) )
            
            buf = []
            state = 0            

ser.close()      

File: src/C#/read_test/Read_test.cs
// Masahiro Furukawa
// July 2, 2020

using System;
using System.IO.Ports;
using System.Linq;
using System.Runtime.CompilerServices;

namespace read_test
{


    class Read_test
    {
        // Coefficient values to translate byte code to real value
        //
        // https://os.mbed.com/compiler/#nav:/MPU-9250-Ch2_ACC_binary/MPU9250_SPI/MPU9250.h
        // /* ---- Sensitivity --------------------------------------------------------- */

        const double MPU9250A_2g = 0.000061035156;       // 0.000061035156 g/LSB
        const double MPU9250A_4g = 0.000122070312;       // 0.000122070312 g/LSB
        const double MPU9250A_8g = 0.000244140625;       // 0.000244140625 g/LSB
        const double MPU9250A_16g = 0.000488281250;      // 0.000488281250 g/LSB

        const double MPU9250G_250dps = 0.007633587786;   // 0.007633587786 dps/LSB
        const double MPU9250G_500dps = 0.015267175572;   // 0.015267175572 dps/LSB
        const double MPU9250G_1000dps = 0.030487804878;  // 0.030487804878 dps/LSB
        const double MPU9250G_2000dps = 0.060975609756;  // 0.060975609756 dps/LSB

        const double MPU9250M_4800uT = 0.6;              // 0.6 uT/LSB

        const double MPU9250T_85degC = 0.002995177763;   // 0.002995177763 degC/LSB
        const double Magnetometer_Sensitivity_Scale_Factor = 0.15;

        static SerialPort serialPort1 = new SerialPort();

        Read_test()
        {
        }

        ~Read_test()
        {
            serialPort1.Close();
        }

        static void Main(string[] args)
        {
            string[] ports = SerialPort.GetPortNames();

            serialPort1.PortName = ports[ports.Count() - 1]; // select the last one
            serialPort1.BaudRate = 921600;
            serialPort1.Parity = Parity.None;
            serialPort1.DataBits = 8;
            serialPort1.StopBits = StopBits.One;
            serialPort1.Handshake = Handshake.None;

            serialPort1.Open();


            if (!serialPort1.IsOpen)
            {
                Console.WriteLine("Could not open COM Port");
                return;
            }

            serialPort1.Write("r"); // command to stop data stream for mbed
            serialPort1.Write("s"); // command to start data stream for mbed

            Int32 val = 0;
            Int32 state = 0;
            int p = 0;
            int count = 0;
            byte[] buf = new byte[51];

            while (true)
            {
                count = serialPort1.BytesToRead;

                while (count > 0)
                {
                    try
                    {
                        val = (byte)serialPort1.ReadByte();
                    }
                    catch
                    {
                        Console.WriteLine("catched exception.  count = {0}", count);
                    }
                    if (state == 0 && val == (byte)'\r')
                    {
                        val = (byte)serialPort1.ReadByte();

                        // waiting for '\n' and get ready
                        if (val == (byte)'\n')
                        {
                            state = 1;
                            p = 0;
                            continue;
                        }
                        else
                        {
                            Console.WriteLine("End byte set failure.");
                        }

                    }

                    // store val into list
                    if (state == 1)
                    {
                        // Make Little Endien for BitConverter from Big Endien
                        if (p % 2 == 0)
                            buf[p + 1] = (byte)val;
                        else
                            buf[p - 1] = (byte)val;
                        p++;

                        // ready to decode
                        if (p == 48)
                        {
                            p = 0;
                            state = 0;

                            short s = 0;
                            for (int i = 0; i < 3; i++) 
                            { 
                                s = BitConverter.ToInt16(buf, 2 * i);
                                Console.Write("{0,5:F} ", s * MPU9250A_4g);
                            }
                            for (int i = 0; i < 3; i++) { 
                                s = BitConverter.ToInt16(buf, 6 + 2 * i); 
                                Console.Write("{0,7:F} ", s * MPU9250G_500dps);
                            }
                            Console.WriteLine();
                        }

                    }
                }
            }


        }

    }
}
